### PR TITLE
[WIP] Remove allow(rustc::potential_query_instability) in rustc_session

### PR DESCRIPTION
**Pull request by NiklasJonsson**
Replace a use of FxHashMap with FxIndexMap

Relates #84447 

Haven't contributed before so not 100 % sure of who to ask for a review but

r? @cjgillot

seems reasonable as you commented last on the issue 😄

### CODE REVIEW

File: compiler/rustc_session/src/parse.rs [hunk 1]
```diff
@@ -56,9 +56,9 @@ impl GatedSpans {
     }
 
     /// Prepend the given set of `spans` onto the set in `self`.
-    pub fn merge(&self, mut spans: FxHashMap<Symbol, Vec<Span>>) {
+    pub fn merge(&self, mut spans: FxIndexMap<Symbol, Vec<Span>>) {
         let mut inner = self.spans.borrow_mut();
-        for (gate, mut gate_spans) in inner.drain() {
+        for (gate, mut gate_spans) in inner.drain(..) {
```
Comment (NiklasJonsson) [inline compiler/rustc_session/src/parse.rs:61]: As this just puts the contents of the drained hashmap into another one, I don't see a way that the iteration order can affect the resulting map. So we could also add a local `allow` here but replacing the map seemed preferable. Though, I'm not sure if this has any performance implications.

**Comment by NiklasJonsson** [comment]
This wasn't as ready as I thought, must've done some mistake with git.

@rustbot label -S-waiting-on-review

**Comment by NiklasJonsson** [comment]
`compiler/rustc_session/src/config.rs` has this code that fails when enabling the lint:
```
impl<T> CheckCfg<T> {
    fn map_data<O: Eq + Hash>(&self, f: impl Fn(&T) -> O) -> CheckCfg<O> {
        CheckCfg {
            names_valid: self
                .names_valid // FxHashSet
                .as_ref()
                .map(|names_valid| names_valid.iter().map(|a| f(a)).collect()),
            values_valid: self
                .values_valid // FxHashMap<T, FxHashSet>
                .iter()
                .map(|(a, b)| (f(a), b.iter().map(|b| f(b)).collect()))
                .collect(),
            well_known_values: self.well_known_values,
        }
    }
}
```
which seems fine enough as it just creates a new hashmap in-place with the mapped values and keys, assuming `f` doesn't have side-effects, which I don't think it can as it is `Fn`. But we want to replace this, which should be straight-forward but the file also has this code:
```
            // Get all values map at once otherwise it would be costly.
            // (8 values * 220 targets ~= 1760 times, at the time of writing this comment).
            let [
                values_target_os,
                values_target_family,
                values_target_arch,
                values_target_endian,
                values_target_env,
                values_target_abi,
                values_target_vendor,
                values_target_pointer_width,
            ] = self
                .values_valid
                .get_many_mut(VALUES)
                .expect("unable to get all the check-cfg values buckets");
```
which makes use of `get_many_mut` which is only implemented on `HashMap` and not the `IndexMap`, which means we can't replace it to avoid the lint. I'll close this for now and see if I can get together something with a smaller scope.

**Comment by jyn514** [comment]
> which makes use of get_many_mut which is only implemented on HashMap and not the IndexMap, which means we can't replace it to avoid the lint

FWIW I think it would be fine to add that API to IndexMap

**Comment by NiklasJonsson** [comment]
> FWIW I think it would be fine to add that API to IndexMap

Yeah I figured but I didn't want to leave this PR in limbo here until that is done (possibly by me).

**Comment by NiklasJonsson** [comment]
`get_many_mut` for `IndexMap` is in a PR [here](https://github.com/bluss/indexmap/pull/238) but blocked in the standard hashmap api getting stabilized.

**Comment by jyn514** [comment]
Cam IndexMap add it under a `nightly` or `unstable` opt-in cargo feature?

**Pull request closed without merging**